addOns/client/src/main/java/org/zaproxy/addon/client/pscan/ClientPassiveScanHelper.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2023 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.addon.client.pscan;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.parosproxy.paros.core.scanner.Alert;
+import org.parosproxy.paros.extension.history.ExtensionHistory;
+import org.parosproxy.paros.model.HistoryReference;
+import org.zaproxy.zap.extension.alert.ExtensionAlert;
+
+public class ClientPassiveScanHelper {
+
+    private static final Logger LOGGER = LogManager.getLogger(ClientPassiveScanHelper.class);
+
+    private ExtensionAlert extAlert;
+    private ExtensionHistory extHistory;
+
+    public ClientPassiveScanHelper(ExtensionAlert extAlert, ExtensionHistory extHistory) {
+        this.extAlert = extAlert;
+        this.extHistory = extHistory;
+    }
+
+    public void raiseAlert(Alert alert) {
+        // Find suitable Href
+        int lastId = extHistory.getLastHistoryId();
+        String alertUrl = alert.getUri();
+        int fragmentOffset = alertUrl.indexOf('#');
+        if (fragmentOffset > 0) {
+            alertUrl = alertUrl.substring(0, fragmentOffset - 1);
+        }
+        // We dont expect to have to go too far back..
+        LOGGER.debug("Searching for href for {}" + alert.getUri());
+        for (int i = lastId; i >= 0; i--) {
+            HistoryReference hr = extHistory.getHistoryReference(i);
+            if (alertUrl.equals(hr.getURI().toString())) {
+                alert.setHistoryRef(hr);
+                LOGGER.debug("Found href {} for {}" + hr.getHistoryId(), alert.getUri());
+                break;
+            }
+        }
+        if (alert.getHistoryRef() == null) {
+            LOGGER.error(
+                    "Failed to find Href for URL {}, unable to raise alert",
```

Comment:
Here and in comments etc above we should say history reference, vs href which might be confused with URL